File: android/core/protocol/src/main/kotlin/app/hapi/protocol/chat/ReducerTools.kt
package app.hapi.protocol.chat

import app.hapi.protocol.wire.AgentState
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull

/** Port of `web/src/chat/reducerTools.ts`. */

data class PermissionEntry(
    val toolName: String,
    /** Verbatim `arguments` (wire default [JsonNull] when absent — projected as `input: null`, like TS `?? null`). */
    val input: JsonElement,
    val permission: ToolPermission,
)

fun getPermissions(agentState: AgentState?): LinkedHashMap<String, PermissionEntry> {
    val map = LinkedHashMap<String, PermissionEntry>()

    val completed = agentState?.completedRequests
    if (completed != null) {
        for ((id, entry) in completed) {
            map[id] = PermissionEntry(
                toolName = entry.tool,
                input = entry.arguments,
                permission = ToolPermission(
                    id = id,
                    status = entry.status,
                    reason = entry.reason,
                    mode = entry.mode,
                    decision = entry.decision,
                    allowedTools = entry.allowTools,
                    answers = entry.answers,
                    createdAt = entry.createdAt,
                    completedAt = entry.completedAt,
                    presence = ToolPermission.COMPLETED_KEYS,
                ),
            )
        }
    }

    val requests = agentState?.requests
    if (requests != null) {
        for ((id, request) in requests) {
            if (map.containsKey(id)) continue
            map[id] = PermissionEntry(
                toolName = request.tool,
                input = request.arguments,
                permission = ToolPermission(
                    id = id,
                    status = "pending",
                    createdAt = request.createdAt,
                    presence = ToolPermission.PENDING_KEYS,
                ),
            )
        }
    }

    return map
}

/** Mutable seed for [ensureToolBlock]; Kotlin null mirrors TS `undefined` throughout. */
class ToolBlockSeed(
    val createdAt: Long,
    val invokedAt: Long? = null,
    val durationMs: Double? = null,
    val usage: UsageData? = null,
    val model: String? = null,
    val localId: String? = null,
    val meta: JsonElement? = null,
    val name: String,
    val input: JsonElement? = null,
    val description: String? = null,
    val nativeTitle: String? = null,
    val nativeKind: String? = null,
    /** Tri-state (TS `progress?: unknown` with an `'in' data` guard at the call site). */
    val hasProgress: Boolean = false,
    val progress: JsonElement? = null,
    val permission: ToolPermission? = null,
    val agentTimestamp: Long? = null,
)

private fun isPlaceholderToolName(name: String): Boolean {
    val normalized = name.trim().lowercase()
    return normalized == "" || normalized == "tool" || normalized == "unknown" || normalized == "generic"
}

fun ensureToolBlock(
    blocks: MutableList<ChatBlock>,
    toolBlocksById: MutableMap<String, ToolCallBlock>,
    id: String,
    seed: ToolBlockSeed,
): ToolCallBlock {
    val existing = toolBlocksById[id]
    if (existing != null) {
        // Preserve earliest createdAt for stable ordering.
        if (seed.createdAt < existing.createdAt) {
            existing.createdAt = seed.createdAt
            existing.tool = existing.tool.copy(createdAt = seed.createdAt)
        }
        if (seed.permission != null) {
            val nextPermission = existing.tool.permission?.mergedWith(seed.permission) ?: seed.permission
            var nextState = existing.tool.state
            if (existing.tool.state == ToolState.RUNNING && seed.permission.status == "pending") {
                nextState = ToolState.PENDING
            }
            existing.tool = existing.tool.copy(permission = nextPermission, state = nextState)
        }
        if (seed.name.isNotEmpty() && (!isPlaceholderToolName(seed.name) || isPlaceholderToolName(existing.tool.name))) {
            existing.tool = existing.tool.copy(name = seed.name)
        }
        if (seed.input != null && seed.input != JsonNull) {
            existing.tool = existing.tool.copy(input = seed.input)
        }
        if (seed.description != null) {
            existing.tool = existing.tool.copy(description = seed.description)
        }
        if (seed.nativeTitle != null) {
            existing.tool = existing.tool.copy(nativeTitle = seed.nativeTitle)
        }
        if (seed.nativeKind != null) {
            existing.tool = existing.tool.copy(nativeKind = seed.nativeKind)
        }
        if (seed.hasProgress && existing.tool.state == ToolState.RUNNING) {
            existing.tool = existing.tool.copy(result = seed.progress ?: JsonNull)
        }
        // tool_use records the invocation time; the tool_result's invokedAt is
        // when the result was processed — keep the original.
        if (seed.invokedAt != null && existing.invokedAt == null) {
            existing.invokedAt = seed.invokedAt
        }
        if (seed.durationMs != null) {
            existing.durationMs = seed.durationMs
        }
        if (seed.usage != null) {
            existing.usage = seed.usage
        }
        if (seed.model != null) {
            existing.model = seed.model
        }
        return existing
    }

    val initialState = when {
        seed.permission?.status == "pending" -> ToolState.PENDING
        seed.permission?.status == "denied" || seed.permission?.status == "canceled" -> ToolState.ERROR
        else -> ToolState.RUNNING
    }

    val tool = ChatToolCall(
        id = id,
        name = seed.name,
        state = initialState,
        input = seed.input,
        createdAt = seed.createdAt,
        startedAt = if (initialState == ToolState.RUNNING) seed.createdAt.toDouble() else null,
        completedAt = null,
        // Exec start only ever comes from a real Claude entry timestamp.
        execStartedAt = if (initialState == ToolState.RUNNING) seed.agentTimestamp?.toDouble() else null,
        execCompletedAt = null,
        description = seed.description,
        nativeTitle = seed.nativeTitle,
        nativeKind = seed.nativeKind,
        result = if (seed.hasProgress) seed.progress ?: JsonNull else null,
        permission = seed.permission,
    )

    val block = ToolCallBlock(
        id = id,
        localId = seed.localId,
        createdAt = seed.createdAt,
        invokedAt = seed.invokedAt,
        durationMs = seed.durationMs,
        usage = seed.usage,
        model = seed.model,
        tool = tool,
        children = emptyList(),
        meta = seed.meta,
    )

    toolBlocksById[id] = block
    blocks.add(block)
    return block
}

fun collectToolIdsFromMessages(messages: List<NormalizedMessage>): Set<String> {
    val ids = mutableSetOf<String>()
    for (msg in messages) {
        val agent = msg as? NormalizedMessage.Agent ?: continue
        for (content in agent.content) {
            when (content) {
                is NormalizedAgentContent.ToolUse -> ids.add(content.id)
                is NormalizedAgentContent.ToolResult -> ids.add(content.toolUseId)
                else -> {}
            }
        }
    }
    return ids
}

fun isChangeTitleToolName(name: String): Boolean =
    name == "mcp__hapi__change_title" || name == "hapi__change_title"

fun extractTitleFromChangeTitleInput(input: JsonElement?): String? {
    val title = asObject(input)?.let { asString(it["title"]) } ?: return null
    return title.trim().takeIf { it.isNotEmpty() }
}

fun collectTitleChanges(messages: List<NormalizedMessage>): Map<String, String> {
    val map = HashMap<String, String>()
    for (msg in messages) {
        val agent = msg as? NormalizedMessage.Agent ?: continue
        for (content in agent.content) {
            val toolUse = content as? NormalizedAgentContent.ToolUse ?: continue
            if (!isChangeTitleToolName(toolUse.name)) continue
            val title = extractTitleFromChangeTitleInput(toolUse.input) ?: continue
            map[toolUse.id] = title
        }
    }
    return map
}

File: ios/Packages/HapiKit/Sources/HapiProtocol/Chat/ReducerTools.swift
import Foundation

// Port of web/src/chat/reducerTools.ts.
//
// The TS reducer mutates block objects shared by reference between the
// timeline array and the by-id map. `BlockBox` reproduces that: a reference
// wrapper around the value-typed `ChatBlock`, mutated in place through the
// settable case accessors (`box.block.asToolCall?.tool.state = …`).

/// Reference cell for one timeline block during reduction. The final
/// pipeline output materializes back to plain `[ChatBlock]`.
final class BlockBox {
    var block: ChatBlock

    init(_ block: ChatBlock) {
        self.block = block
    }
}

extension Array where Element == BlockBox {
    var materialized: [ChatBlock] {
        map(\.block)
    }
}

/// Port of `PermissionEntry`.
struct PermissionEntry {
    var toolName: String
    /// `nil` == the request carried no arguments (TS `undefined`).
    var input: JSONValue?
    var permission: ToolPermission
}

/// Ordered stand-in for the TS `Map<string, PermissionEntry>`.
///
/// The reference iterates JS object keys in insertion order; fixture inputs
/// are canonically serialized (recursively sorted keys) and `JSON.parse`
/// preserves that order, so the order the reference actually observes when
/// replaying fixtures is ascending key order — which is what this map
/// imposes deterministically (completed requests first, then pending,
/// each sorted by request id).
struct PermissionMap {
    private(set) var order: [String] = []
    private(set) var byId: [String: PermissionEntry] = [:]

    subscript(id: String) -> PermissionEntry? {
        byId[id]
    }

    func has(_ id: String) -> Bool {
        byId[id] != nil
    }

    mutating func set(_ id: String, _ entry: PermissionEntry) {
        if byId[id] == nil { order.append(id) }
        byId[id] = entry
    }

    var entriesInOrder: [(id: String, entry: PermissionEntry)] {
        order.map { ($0, byId[$0]!) }
    }
}

/// Port of `getPermissions` — completed requests first (they win the id),
/// then pending requests.
func getPermissions(_ agentState: AgentState?) -> PermissionMap {
    var map = PermissionMap()

    if let completed = agentState?.completedRequests {
        for id in completed.keys.sorted(by: { $0.utf16.lexicographicallyPrecedes($1.utf16) }) {
            let entry = completed[id]!
            let status: ToolPermissionStatus = {
                switch entry.status {
                case .canceled: return .canceled
                case .denied: return .denied
                case .approved: return .approved
                }
            }()
            map.set(id, PermissionEntry(
                toolName: entry.tool,
                input: entry.arguments,
                permission: ToolPermission(
                    id: id,
                    status: status,
                    reason: entry.reason,
                    mode: entry.mode,
                    allowedTools: entry.allowTools,
                    decision: entry.decision,
                    answers: entry.answers,
                    createdAt: entry.createdAt,
                    completedAt: entry.completedAt,
                    // Keys the TS object literal carries (spread overwrites
                    // these on merge even when the value is undefined);
                    // `date` is absent from this construction site.
                    presentKeys: [.reason, .mode, .decision, .allowedTools, .answers, .createdAt, .completedAt]
                )
            ))
        }
    }

    if let requests = agentState?.requests {
        for id in requests.keys.sorted(by: { $0.utf16.lexicographicallyPrecedes($1.utf16) }) {
            if map.has(id) { continue }
            let request = requests[id]!
            map.set(id, PermissionEntry(
                toolName: request.tool,
                input: request.arguments,
                permission: ToolPermission(
                    id: id,
                    status: .pending,
                    createdAt: request.createdAt,
                    presentKeys: [.createdAt]
                )
            ))
        }
    }

    return map
}

/// Seed for `ensureToolBlock` — mirrors the TS parameter object. Optional
/// fields left `nil` reproduce an absent/`undefined` TS property.
struct ToolBlockSeed {
    var createdAt: Int
    var invokedAt: Int?
    var durationMs: Int?
    var usage: UsageData?
    var model: String?
    var localId: String?
    var meta: JSONValue?
    var name: String
    var input: JSONValue?
    var description: String?
    var nativeTitle: String?
    var nativeKind: String?
    var progress: JSONValue?
    var permission: ToolPermission?
    var agentTimestamp: Int?

    init(
        createdAt: Int,
        invokedAt: Int? = nil,
        durationMs: Int? = nil,
        usage: UsageData? = nil,
        model: String? = nil,
        localId: String? = nil,
        meta: JSONValue? = nil,
        name: String,
        input: JSONValue? = nil,
        description: String? = nil,
        nativeTitle: String? = nil,
        nativeKind: String? = nil,
        progress: JSONValue? = nil,
        permission: ToolPermission? = nil,
        agentTimestamp: Int? = nil
    ) {
        self.createdAt = createdAt
        self.invokedAt = invokedAt
        self.durationMs = durationMs
        self.usage = usage
        self.model = model
        self.localId = localId
        self.meta = meta
        self.name = name
        self.input = input
        self.description = description
        self.nativeTitle = nativeTitle
        self.nativeKind = nativeKind
        self.progress = progress
        self.permission = permission
        self.agentTimestamp = agentTimestamp
    }
}

private func isPlaceholderToolName(_ name: String) -> Bool {
    let normalized = name.jsTrimmed.lowercased()
    // 'generic' is agy's non-specific result wrapper — never let it
    // overwrite a specific name set by the permission request.
    return normalized.isEmpty || normalized == "tool" || normalized == "unknown" || normalized == "generic"
}

/// Port of `ensureToolBlock`: create-or-merge the tool card for `id`.
/// Mutates the shared box in place; appends a new box to `blocks` (and the
/// map) when the id is unseen.
@discardableResult
func ensureToolBlock(
    _ blocks: inout [BlockBox],
    _ toolBlocksById: inout [String: BlockBox],
    id: String,
    seed: ToolBlockSeed
) -> BlockBox {
    if let existingBox = toolBlocksById[id], var existing = existingBox.block.asToolCall {
        // Preserve earliest createdAt for stable ordering.
        if seed.createdAt < existing.createdAt {
            existing.createdAt = seed.createdAt
            existing.tool.createdAt = seed.createdAt
        }
        if let seedPermission = seed.permission {
            let nextPermission = ToolPermission.spreading(existing.tool.permission, with: seedPermission)
            var nextState = existing.tool.state
            if existing.tool.state == .running && seedPermission.status == .pending {
                nextState = .pending
            }
            existing.tool.permission = nextPermission
            existing.tool.state = nextState
        }
        if !seed.name.isEmpty && (!isPlaceholderToolName(seed.name) || isPlaceholderToolName(existing.tool.name)) {
            existing.tool.name = seed.name
        }
        if let input = seed.input, input != .null {
            existing.tool.input = input
        }
        if let description = seed.description {
            existing.tool.description = description
        }
        if let nativeTitle = seed.nativeTitle {
            existing.tool.nativeTitle = nativeTitle
        }
        if let nativeKind = seed.nativeKind {
            existing.tool.nativeKind = nativeKind
        }
        if let progress = seed.progress, existing.tool.state == .running {
            existing.tool.result = progress
        }
        // The tool_use records when the tool was invoked; a later
        // tool_result must not overwrite it.
        if let invokedAt = seed.invokedAt, existing.invokedAt == nil {
            existing.invokedAt = invokedAt
        }
        if let durationMs = seed.durationMs {
            existing.durationMs = durationMs
        }
        if let usage = seed.usage {
            existing.usage = usage
        }
        if let model = seed.model {
            existing.model = model
        }
        existingBox.block = .toolCall(existing)
        return existingBox
    }

    let initialState: ToolCallState = {
        if seed.permission?.status == .pending { return .pending }
        if seed.permission?.status == .denied || seed.permission?.status == .canceled { return .error }
        return .running
    }()

    let tool = ChatToolCall(
        id: id,
        name: seed.name,
        state: initialState,
        input: seed.input,
        createdAt: seed.createdAt,
        startedAt: initialState == .running ? seed.createdAt : nil,
        completedAt: nil,
        // Exec start is only ever a real Claude entry timestamp (never the
        // hub receive time); nil keeps the hub-time fallback.
        execStartedAt: initialState == .running ? seed.agentTimestamp : nil,
        execCompletedAt: nil,
        description: seed.description,
        nativeTitle: seed.nativeTitle,
        nativeKind: seed.nativeKind,
        result: seed.progress,
        permission: seed.permission
    )

    let block = ToolCallBlock(
        id: id,
        localId: seed.localId,
        createdAt: seed.createdAt,
        invokedAt: seed.invokedAt,
        durationMs: seed.durationMs,
        usage: seed.usage,
        model: seed.model,
        tool: tool,
        children: [],
        meta: seed.meta
    )

    let box = BlockBox(.toolCall(block))
    toolBlocksById[id] = box
    blocks.append(box)
    return box
}

/// Port of `collectToolIdsFromMessages`.
func collectToolIdsFromMessages(_ messages: [NormalizedMessage]) -> Set<String> {
    var ids = Set<String>()
    for msg in messages {
        guard let content = msg.agentContent else { continue }
        for item in content {
            switch item {
            case .toolCall(let toolCall):
                ids.insert(toolCall.id)
            case .toolResult(let toolResult):
                ids.insert(toolResult.toolUseId)
            default:
                break
            }
        }
    }
    return ids
}

/// Port of `isChangeTitleToolName`.
func isChangeTitleToolName(_ name: String) -> Bool {
    name == "mcp__hapi__change_title" || name == "hapi__change_title"
}

/// Port of `extractTitleFromChangeTitleInput`.
func extractTitleFromChangeTitleInput(_ input: JSONValue?) -> String? {
    guard let title = input?.objectValue?["title"]?.stringValue else { return nil }
    let trimmed = title.jsTrimmed
    return trimmed.isEmpty ? nil : trimmed
}

/// Port of `collectTitleChanges` — tool_use id → requested title.
func collectTitleChanges(_ messages: [NormalizedMessage]) -> [String: String] {
    var map: [String: String] = [:]
    for msg in messages {
        guard let content = msg.agentContent else { continue }
        for item in content {
            guard case .toolCall(let toolCall) = item else { continue }
            guard isChangeTitleToolName(toolCall.name) else { continue }
            guard let title = extractTitleFromChangeTitleInput(toolCall.input) else { continue }
            map[toolCall.id] = title
        }
    }
    return map
}
